### PR TITLE
feat(auth): standalone enable oauth listener for MPAs

### PR DESCRIPTION
**Pull request by HuiSF**
#### Description of changes

This PR introduces an importable and standalone side effect for enabling the OAuth listener for `signInWithRedirect` API in multipage app use cases.

Developers should be able to `import 'aws-amplify/auth/enable-oauth-listener'` in a route (as the `signInWithRedirect` redirect url) to complete the OAuth sign in flow.

* Refactored the original `signInWithRedirect` implementation
  * Modularized the implementation to make the side effect part can be used independently
  * Mostly copy out and paste into new files
  * Fixed issues in the original impl. caught by writing new unit tests
* Refactored the way how the listener is being triggered
  * Attached the listener to the `Amplify` singleton where whether the singleton has a valid `OAuth` config is the single trigger of the listener - this prevents the listener to be executed multiple times on repetitive calls of `Amplify.configure` when someone tries manual configurations
* Refactored the auth unit tests suite
  * To avoid enabling the side effect in irrelevant tests
  * Rewrote unit tests for `signInWithRedirect` and side effect implementation

#### Description of how you validated changes

* Unit tests
* Local tests with
  * Next.js sample app
  * react-native sample app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/utils/oauth/inflightPromiseManager.ts
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/utils/oauth/inflightPromiseManager.ts:1]: super nit: `XXManager` naming is OOD, `inflightPromise.ts` is sufficient IMO.

File: packages/core/src/singleton/Amplify.ts [hunk 1]
```diff
@@ -78,6 +89,29 @@ export class AmplifyClass {
...
+	private notifyOAuthListener() {
+		if (
+			!this.resourcesConfig.Auth?.Cognito.loginWith?.oauth ||
+			!this.oAuthListener
+		) {
+			return;
+		}
+
+		this.oAuthListener(this.resourcesConfig.Auth?.Cognito);
+		this.oAuthListener = undefined;
```
Comment (AllanZhengYP) [inline packages/core/src/singleton/Amplify.ts:113]: It's better to add some explanations for what it does.

File: packages/aws-amplify/src/auth/enableOAuthListener.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// import * as enableOAuthListener from '@aws-amplify/auth/enable-oauth-listener';
+
+// export default enableOAuthListener;
```
Comment (AllanZhengYP) [inline packages/aws-amplify/src/auth/enableOAuthListener.ts:6]: Extra lines?

File: packages/core/src/singleton/Amplify.ts [hunk 3]
```diff
@@ -78,6 +89,29 @@ export class AmplifyClass {
 	getConfig(): Readonly<ResourcesConfig> {
 		return this.resourcesConfig;
 	}
+
+	/** @internal */
+	[ADD_OAUTH_LISTENER](listener: (authConfig: AuthConfig['Cognito']) => void) {
+		if (this.resourcesConfig.Auth?.Cognito.loginWith?.oauth) {
```
Comment (HuiSF) [inline packages/core/src/singleton/Amplify.ts:95]: Didn't use the assertion function here to reduce bundle size impact...

File: packages/core/src/singleton/Amplify.ts [hunk 4]
```diff
@@ -78,6 +89,29 @@ export class AmplifyClass {
...
+			listener(this.resourcesConfig.Auth?.Cognito);
+		} else {
+			// otherwise register the listener and run it later when Amplify gets configured with a valid oauth config
+			this.oAuthListener = listener;
+		}
+	}
+
+	private notifyOAuthListener() {
+		if (
+			!this.resourcesConfig.Auth?.Cognito.loginWith?.oauth ||
```
Comment (HuiSF) [inline packages/core/src/singleton/Amplify.ts:106]: Ditto, Bundle size impact.

**Review by AllanZhengYP — COMMENTED** [review]
It looks good! Thank you @HuiSF

**Review by AllanZhengYP — APPROVED** [review]

File: packages/auth/src/providers/cognito/utils/oauth/attemptCompleteOAuthFlow.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+			currentUrl,
+			clientId: userPoolClientId,
+			domain,
+			redirectUri,
+			responseType,
+			userAgentValue: getAuthUserAgentValue(AuthAction.SignInWithRedirect),
+		});
+	} catch (err) {
+		// no-op
+		// Possible errors:
```
Comment (israx) [inline packages/auth/src/providers/cognito/utils/oauth/attemptCompleteOAuthFlow.ts:50]: if `getRedirectUrl` throws, then it is not possible to surface the error. However this is unlikely, the `getRedirectUrl` method is invoked before starting an oauth flow.

Comment (HuiSF) [reply packages/auth/src/providers/cognito/utils/oauth/attemptCompleteOAuthFlow.ts:50]: Good observation.

Probably need to handle the error may be thrown from `const redirectUri = getRedirectUrl(redirectSignIn);`. When the error happens it should reset the oauth flow state? Actually, all the errors may be caught in the `attemptCompleteOAuthFlow` should trigger resetting the oauth flow? wdyt?

**Review by israx — APPROVED** [review]
LGTM! I smoked test these changes with a React App and works fine!

**Review by israx — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Review by kvramyasri7 — APPROVED** [review]

**Review by Ashish-Nanda — APPROVED** [review]
Approved. Lets release to the preview tag first for further testing and validation of this change.

**Pull request merged**